File: Java/architecture-patterns/src/main/java/example/repository/OrderRepository.java
package example.repository;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

/**
 * OrderRepository — DATA ACCESS LAYER (compliant)
 *
 * Responsibilities:
 *   ✓ Owns all DB/storage interaction
 *   ✓ No imports from controller or service packages
 *   ✓ Exposes a clean interface consumed by the service layer only
 */
public class OrderRepository {

    private final Map<String, Order> store = new HashMap<>();

    public Optional<Order> findById(String orderId) {
        return Optional.ofNullable(store.get(orderId));
    }

    public Order save(Order order) {
        store.put(order.orderId(), order);
        return order;
    }

    public record Order(String orderId, String userId, double amount, String status) {}
}

File: dotnet/architecture-patterns/src/Repositories/OrderRepository.cs
namespace ArchitecturePatterns.Repositories;

/// <summary>
/// OrderRepository — DATA ACCESS LAYER (compliant)
///
/// Architecture rules enforced here:
///   ✓ No using directives from Controllers or Services namespaces
///   ✓ Owns all storage interaction
///   ✓ Implements IOrderRepository interface
/// </summary>
public class OrderRepository : IOrderRepository
{
    private readonly Dictionary<string, Order> _store = new();

    public Order? FindById(string orderId) =>
        _store.TryGetValue(orderId, out var order) ? order : null;

    public Order Save(Order order)
    {
        _store[order.OrderId] = order;
        return order;
    }
}
